src/btree/iter.rs

```diff
@@ -344,79 +353,121 @@ impl BTreeIterState {
 		btree: &mut BTree,
 		col: &BTreeTable,
 		log: &impl LogQuery,
+		direction: IterDirection,
 	) -> Result<Option<(Vec<u8>, Value)>> {
 		if self.fetch_root {
 			let root = col.with_locked(|tables| {
 				BTree::fetch_root(btree.root_index.unwrap_or(NULL_ADDRESS), tables, log)
 			})?;
-			self.state.push((0, NodeType::Child, root));
+			self.state.push((direction.starting(), root));
 			self.fetch_root = false;
 		}
 
-		while !self.state.is_empty() {
-			while let Some((ix, ty @ NodeType::Child, node)) = self.state.last_mut() {
-				*ty = NodeType::Separator;
-				if let Some(child) = col.with_locked(|btree| node.fetch_child(*ix, btree, log))? {
-					self.state.push((0, NodeType::Child, child));
-				}
-			}
-
-			let (ix, ty, node) = self.state.last_mut().expect("We always have at least one entry");
-			if *ix < ORDER {
-				if let Some(address) = node.separator_address(*ix) {
-					let key = node.separator_key(*ix).unwrap();
-					*ix += 1;
-					*ty = NodeType::Child;
-
-					let key_query = TableKeyQuery::Fetch(None);
-					let r = col.get_at_value_index(key_query, address, log)?;
-					return Ok(r.map(|r| (key, r.1)))
-				}
-			}
-
-			self.state.pop();
-		}
-
-		Ok(None)
-	}
-
-	pub fn prev(
-		&mut self,
-		col: &BTreeTable,
-		log: &impl LogQuery,
-	) -> Result<Option<(Vec<u8>, Value)>> {
-		while !self.state.is_empty() {
-			while let Some((ix, ty @ NodeType::Child, node)) = self.state.last_mut() {
-				*ty = NodeType::Separator;
-				match col.with_locked(|btree| node.fetch_child(*ix, btree, log))? {
-					Some(child) => {
-						let last_separator = child.last_separator_index().unwrap_or_default();
-						let last_child = child.last_child_index().unwrap_or_default();
-						if *ix > 0 {
-							*ix -= 1;
+		loop {
+			let is_leaf = btree.depth as usize + 1 == self.state.len();
+			if let Some(state) = self.state.last_mut() {
+				let next = match (direction, &state.0) {
+					(_, LastIndex::Descend(_)) => unreachable!("exit function clean it"),
```

Comment:
```suggestion
					(_, LastIndex::Descend(_)) => unreachable!("Las state should be clean after calling self.exit"),
```

Is it always cleaned by `self.exit` though? From what I can follow, sometimes `self.enter` is called instead.